pkg/reconcilermanager/controllers/reposync_controller_test.go

```diff
@@ -2532,6 +2534,67 @@ func TestRepoSyncReconcileStaleClientCache(t *testing.T) {
 	require.Contains(t, reconcilingCondition.Message, "RepoSyncs must specify spec.git when spec.sourceType is \"git\"", "unexpected Stalled condition message")
 }
 
+func TestUpdateReconcilerWithAllowVerticalScale(t *testing.T) {
+	// Mock out parseDeployment for testing.
+	parseDeployment = parsedDeployment
+
+	rs := repoSync(reposyncNs, reposyncName, reposyncRef(gitRevision), reposyncBranch(branch), reposyncSecretType(configsync.AuthSSH), reposyncSecretRef(reposyncSSHKey))
+	reqNamespacedName := namespacedName(rs.Name, rs.Namespace)
+	// mock allow vertical scale setting
+	allowVerticalScale = true
+	fakeClient, testReconciler := setupNSReconciler(t, rs, secretObj(t, reposyncSSHKey, configsync.AuthSSH, v1beta1.GitSource, core.Namespace(rs.Namespace)))
+
+	// Test creating Deployment resources.
+	ctx := context.Background()
+	if _, err := testReconciler.Reconcile(ctx, reqNamespacedName); err != nil {
+		t.Fatalf("unexpected reconciliation error, got error: %q, want error: nil", err)
+	}
+	repoDeployment := repoSyncDeployment(
+		nsReconcilerName,
+		setServiceAccountName(nsReconcilerName),
+	)
+
+	updatedCPU := "249m"
+	updatedMemory := "259Mi"
+	deploymentCoreObject := fakeClient.Objects[core.IDOf(repoDeployment)]
+	wantDeployments := deploymentCoreObject.(*appsv1.Deployment)
+	initialReplica := *wantDeployments.Spec.Replicas
+	var updatedReplica int32 = 20
+	*wantDeployments.Spec.Replicas = updatedReplica
+	// mock VPA behavior of changing deployment resources
+	wantDeployments.Spec.Template.Spec.Containers[0].Resources.Requests = corev1.ResourceList{
+		corev1.ResourceCPU:    resource.MustParse(updatedCPU),
+		corev1.ResourceMemory: resource.MustParse(updatedMemory),
+	}
+
+	if err := fakeClient.Update(ctx, wantDeployments); err != nil {
+		t.Fatalf("failed to update the repo sync request, got error: %v", err)
+	}
+	if _, err := testReconciler.Reconcile(ctx, reqNamespacedName); err != nil {
+		t.Fatalf("unexpected reconciliation error upon request update, got error: %q, want error: nil", err)
+	}
+
+	// when vertical scale is enabled, original deployment resource should not be updated
+	deploymentCoreObject = fakeClient.Objects[core.IDOf(repoDeployment)]
+	wantDeployments = deploymentCoreObject.(*appsv1.Deployment)
+	for _, container := range wantDeployments.Spec.Template.Spec.Containers {
+		if container.Name == reconcilermanager.Reconciler {
+			currentCPU := container.Resources.Requests.Cpu().String()
+			if updatedCPU != currentCPU {
+				t.Errorf("Deployment validation failed. Expected CPU: %s, got %s", updatedCPU, currentCPU)
```

Comment:
Please reuse the `validateDeployments` function to verify.

Reply:
validation has been modified to construct expected deployment and using validateDeployments function